### ISSUE TITLE
Model loading give different results on Windows and Linux

### ISSUE THREAD

**Issue by jolinxql** [pytorch/pytorch#4555]
I found that when I load and evaluate models trained and saved on windows use different operating systems, results are **inconsistent**.

**Phenomenon:**
I trained a model on Windows 10 and tested it on the test set before saving and got `accuracy = 63.91%`. Then I saved it to "model_epoN". 
1. When I load and evaluate it on the same Windows 10, the results are **consistent**. i.e., `accuracy = 63.91%`, too.
2. But when I load and evaluate it on Ubuntu, the results are **inconsistent**. `accuracy = 0.1%`, like randomly guessing.

I've noticed the difference between CR LF and LF line break types and made small changes to the datasets, so I'm sure the problem lies somewhere else.
Please note: Different to Issue https://github.com/pytorch/pytorch/issues/4333 whose difference is small, the difference I got is huge.

**Version information:**
The pytorch version on Windows 10 python3.6 is `0.3.0b0+591e73e` (peterjc123).
The pytorch version on Ubuntu 16.4 python3.5 is `0.3.0.post4` (official).

**Save function:**
```
def save_checkpoint(parser, epoch):
    torch.save({'state_dict': parser.state_dict()},
    'model_epo' + str(epoch + 1))
```
**Load function:**
```
def load_checkpoint(filename, parser):
    checkpoint = torch.load(filename)
    parser.load_state_dict(checkpoint['state_dict'])
    return parser
```
No error or warning was raised. I guess the formats for saving state_dict are different according to the OS? 
Currently, I'm not able to write a minimal working example because the model is too complicated and it builds dynamic nets in each iteration. I haven't trained the model on Linux and test it on Win. But I will try and I still need some time.
I hope the problem can be solved in pytorch v0.4.0.

**Comment by peterjc123** [comment]
@jolinxql According to the post [here](https://stackoverflow.com/questions/1849523/is-pickle-file-of-python-cross-platform), It seems that Pickle is not consistent when you load dumps from a newer version in a old version of Pickle. Would you please try whether it is okay when the two PCs share the same Python version?

**Comment by jolinxql** [comment]
@peterjc123 Thank you for your swift reply! I created a python 3.6 conda env on Ubuntu and tested. But the accuracy still equals to 0.1%. The output of python3.5 and python3.6 both on ubuntu are exactly same. 
Maybe I should also try another test on python 3.5+win10.

**Comment by ezyang** [comment]
If accuracy is 0.1%, that means that some part of the network is behaving completely differently, which means it should be easy to debug layer by layer on a specific input and see where the big change is happening. Is that possible? Try deleting layers from the end of your model until the resulting tensors match, and then the problematic layer will be the one.

**Comment by jolinxql** [comment]
@ezyang Ok, I agree. I'll create a new post when I figure it out.

**Comment by hfutyanhuan** [comment]
Have you solved it？ @jolinxql

**Comment by jolinxql** [comment]
> Have you solved it？ @jolinxql

I found following reasons

1. Pytorch's CPU's and GPU's seeds are independent. So set/save/load them independently

https://discuss.pytorch.org/t/are-gpu-and-cpu-random-seeds-independent/142

2. If you use randomness on severall gpus, you need to set torch.cuda.manual_seed_all(seed).

If you use cudnn, you need to set torch.backends.cudnn.deterministic=True.

3. There are many lines that set random seeds in different functions. E.g.,

linear_layer_init(seed=1)

sample_next_word(seed=10)

4. Different random generators

import random as rng

import numpy.random as rng

import torch; torch.cuda.manual_seed()

5. Forget to save random seed before testing the model on test/dev sets (which also do sampling or shuffle or simply setting seeds)

6. Forget to test immediately after loading the model before resuming training (because testing(random sample involved, performance showed) is always at the end of each epoch)
